test/src/test/integration/git_protocol.rs

```diff
@@ -0,0 +1,384 @@
+
+    clone_with(&remote, &local, move |stop| {
+        packwriter::Libgit::new(packwriter::Options::default(), local_repo, stop)
+    })
+}
+
+#[test]
+fn clone_gitoxide() {
+    let remote = upstream();
+    let local = tempdir().unwrap();
+    let local_repo = git2::Repository::init(&local).unwrap();
+
+    clone_with(&remote, &local, move |stop| {
+        packwriter::Standard::new(local_repo.path(), packwriter::Options::default(), stop)
+    })
+}
+
+fn thin_pack_with<R, L, B, P>(remote: R, local: L, build_pack_writer: B)
+where
+    R: AsRef<Path>,
+    L: AsRef<Path>,
+    B: Fn(Arc<AtomicBool>) -> P,
+    P: PackWriter + Send + 'static,
+    P::Output: Send + 'static,
+{
+    // Clone main only
+    {
+        let (client_out, server_out) = run_fetch(
+            &remote,
+            fetch::Options {
+                repo: "foo".into(),
+                extra_params: vec![],
+                ref_prefixes: None,
+                want_refs: vec![WantRef {
+                    name: "refs/heads/main".into(),
+                    have: None,
+                }],
+                done_after_pack: true,
+            },
+            want_id,
+            &build_pack_writer,
+        )
+        .unwrap();
+
+        assert!(server_out.success());
+        assert!(client_out.pack.is_some());
+    }
+
+    let remote_repo = git2::Repository::open(&remote).unwrap();
+    remote_repo.set_namespace("foo").unwrap();
+    let local_repo = git2::Repository::open(&local).unwrap();
+
+    // Fetch next, which is ahead of main
+    {
+        let head = remote_repo.refname_to_id("refs/heads/main").unwrap();
+        let (client_out, server_out) = run_fetch(
+            &remote,
+            fetch::Options {
+                repo: "foo".into(),
+                extra_params: vec![],
+                ref_prefixes: None,
+                want_refs: vec![WantRef {
+                    name: "refs/heads/next".into(),
+                    have: Some(ObjectId::from_20_bytes(head.as_bytes())),
+                }],
+                done_after_pack: true,
+            },
+            want_id,
+            build_pack_writer,
+        )
+        .unwrap();
+        assert!(server_out.success());
+        assert!(client_out.pack.is_some());
+
+        update_tips(&local_repo, &client_out.refs).unwrap();
+    }
+
+    let remote_history = collect_history(&remote_repo, "refs/heads/next").unwrap();
+    let local_history = collect_history(&local_repo, "refs/heads/next").unwrap();
+
+    assert!(!remote_history.is_empty());
+    assert_eq!(remote_history, local_history)
+}
+
+#[test]
+fn thin_pack_libgit() {
+    let remote = upstream();
+    let local = tempdir().unwrap();
+
+    thin_pack_with(&remote, &local, |stop| {
+        let local_repo = git2::Repository::init(&local).unwrap();
+        packwriter::Libgit::new(packwriter::Options::default(), local_repo, stop)
+    })
+}
+
+#[test]
+fn thin_pack_gitoxide() {
```

Comment:
I am definitely loving these tests and am glad it does indeed work as expected.
Currently thin pack support is tested on a thin pack of `gitoxide`'s own making _only_.